prog/dftb+/lib_extlibs/negf_int.F90

```diff
@@ -0,0 +1,2117 @@
+  ! using negf_init_csr 
+  public :: negf_init_csr
+  type(z_CSR) :: csrHam, csrOver
+
+  ! non wrapped direct calls  
+  private :: negf_density, negf_current, negf_ldos
+
+  contains
+
+  !------------------------------------------------------------------------------
+  ! Init gDFTB environment and variables
+  !
+  ! Note: mpicomm should be the global commworld here
+  !------------------------------------------------------------------------------
+  subroutine negf_init(transpar, greendens, tundos, mpicomm,&
+        & tempElec, initinfo)
+    
+    Type(TTranspar), intent(IN) :: transpar
+    Type(TNEGFGreenDensInfo), intent(IN) :: greendens
+    Type(TNEGFTunDos), intent(IN) :: tundos
+    Type(mpifx_comm), intent(in) :: mpicomm
+    real(dp), intent(in) :: tempElec 
+    logical, intent(OUT) :: initinfo
+    
+    ! local variables
+    real(dp), dimension(:), allocatable :: pot, eFermi
+    integer :: i,error, l, ncont, nc_vec(1), j, nldos
+    integer, dimension(:), allocatable :: sizes  
+    ! string needed to hold processor name
+    character(:), allocatable :: hostname
+    type(lnParams) :: params
+   
+    error = 0 
+    initinfo = .true.       
+
+    call negf_mpi_init(mpicomm)
+     
+    if (transpar%defined) then
+      ncont = transpar%ncont 
+    else
+      ncont = 0
+    endif
+    
+    ! ------------------------------------------------------------------------------
+    !! Set defaults and fill up the parameter structure with them
+    call init_negf(negf)
+    call init_contacts(negf, ncont)
+    call set_scratch(negf, ".")
+    if (id0) call create_scratch(negf)    
+    call get_params(negf, params)
+    
+    ! ------------------------------------------------------------------------------
+    !This must be different for different initialisations, to be separated
+    !Higher between transport and greendens is taken, temporary
+    if (tundos%defined .and. greendens%defined) then                                          
+       if (tundos%verbose.gt.greendens%verbose) then 
+          params%verbose = tundos%verbose
+       else
+          params%verbose = greendens%verbose
+       endif
+    else
+       if (tundos%defined) params%verbose = tundos%verbose
+       if (greendens%defined) params%verbose = greendens%verbose
+    end if
+    ! ------------------------------------------------------------------------------
+    ! This parameter is used to set the averall drop threshold in libnegf
+    ! It affects especially transmission that is not accurate more than 
+    ! this value.
+    call set_drop(1.d-20)
+
+ 
+    ! ------------------------------------------------------------------------------
+    ! Assign spin degenracy and check consistency between different input blocks
+    if (tundos%defined .and. greendens%defined) then
+      if (tundos%gSpin .ne. greendens%gSpin) then
+        write(*,*) "ERROR: spin degeneracy is not consistent between different input blocks"
```

Comment:
Write(*, statements won't connect with the parallel write-out structure of the main code - things like should this be appearing on rank=0 or from anything in the comm world?